noir/noir-repo/compiler/noirc_frontend/src/monomorphization/mod.rs

```diff
@@ -314,9 +329,20 @@ impl<'interner> Monomorphizer<'interner> {
         });
         let unconstrained = modifiers.is_unconstrained;
 
+        let should_fold = meta.should_fold;
+
         let parameters = self.parameters(&meta.parameters);
         let body = self.expr(body_expr_id)?;
-        let function = ast::Function { id, name, parameters, body, return_type, unconstrained };
+        let function = ast::Function {
+            id,
+            name,
+            parameters,
+            body,
+            return_type,
+            unconstrained,
+            should_fold,
+            func_sig,
+        };
```

Comment:
I think we're missing something in type checking which enforces that a foldable function's signature is a valid entrypoint.

Reply:
I changed it so that we check the program inputs in type checking. I could generalize it further to state a foldable function returns true for `is_entry_point_function`. This would then require the outputs of all foldable functions to be marked with `pub`. I didn't generalize to have foldable functions return true for `is_entry_point_function` due to some of the discussions we have had about private return values as well as it is not super clear how the outputs of foldable functions should be treated. They most likely won't be true "public inputs", but instead live as part of the databus.

Reply:
I've included a test in type check for this